### PR TITLE
refactor: convert puppeteer to TS

### PR DESCRIPTION
**Pull request by vladfrangu**
CC: @pocesar @B4nan @mnmkng

This PR also updates TS, and disables npm 7's stupid peer dependency changes

### CODE REVIEW

**Comment by vladfrangu** [comment]
CI is failing because it still has the old TS version it seems... anything I can do to fix that? 😓

File: .npmrc [hunk 1]
```diff
@@ -0,0 +1 @@
+legacy-peer-deps=true
```
Comment (szmarczak) [inline .npmrc:1]: Why is this needed?

Comment (vladfrangu) [reply .npmrc:1]: Was added because I tried to update typescript here, yet the eslint rules use an older version (and npm 7 threw errors due to its new peer dep parsing which basically made peerDeps a non-optional-optionalDependencies)

Personally, I use this everywhere (I much prefer npm <7's behavior), but I don't know if it's wanted here too 👀 Some input would be lovely!

File: src/launch-context.ts [hunk 2]
```diff
@@ -67,6 +67,9 @@ export class LaunchContext<
 
     private readonly _reservedFieldNames = [...Reflect.ownKeys(this), 'extend'];
 
+    // TODO: change this to PropertyKey
```
Comment (szmarczak) [inline src/launch-context.ts:70]:
```suggestion
    // TODO: change this to PropertyKey when TypeScript 4.4 releases
```

https://github.com/microsoft/TypeScript/pull/44512 :)

File: .npmrc (hunk 1 above)
Comment (szmarczak) [reply .npmrc:1]: https://github.com/typescript-eslint/typescript-eslint/blob/v4.29.0/package.json#L124
https://cdn.jsdelivr.net/npm/@apify/eslint-config-ts@0.1.2/package.json

Hmm... I removed `legacy-peer-deps` and it installed just fine. Am I missing something?

Comment (szmarczak) [reply .npmrc:1]: > Personally, I use this everywhere (I much prefer npm <7's behavior), but I don't know if it's wanted here too eyes

I have never used this so I can't say what's better. I guess it's ok as long as it works, but not sure.

Comment (vladfrangu) [reply .npmrc:1]: Oh, right, I was installing `@next` since that was the version that had the fix in, that'd explain why 😅 

Yeah, I can remove this if wanted 👍

**Review by szmarczak — APPROVED** [review]

File: .npmrc (hunk 1 above)
Comment (mnmkng) [reply .npmrc:1]: Let's keep it for now. It's safer. We can make it more strict later.
